### PR TITLE
[libcasr] cargo-fuzz support

### PR DESCRIPTION
**Pull request by anfedotoff**

### CODE REVIEW

File: libcasr/fuzz/fuzz_targets/parse_stacktrace.rs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+use libcasr::stacktrace::ParseStacktrace;
+
+fuzz_target!(|data: &[u8]| {
+    if data.len() < 2 {
+        return;
+    }
+    let s = String::from_utf8_lossy(&data[1..]);
+    match data[0] % 5 {
+        0 => {
+            // Asan
```
Comment (SweetVishnya) [inline libcasr/fuzz/fuzz_targets/parse_stacktrace.rs:19]: Add UBSAN?

File: libcasr/Cargo.toml [hunk 2]
```diff
@@ -9,7 +9,7 @@ repository = "https://github.com/ispras/casr.git"
 documentation = "https://docs.rs/libcasr"
 edition = "2021"
 license = "Apache-2.0"
-exclude = ["/tests"]
```
Comment (SweetVishnya) [inline libcasr/Cargo.toml:12]: exclude both tests and fuzz

File: libcasr/fuzz/fuzz_targets/parse_stacktrace.rs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+use libfuzzer_sys::fuzz_target;
+
+use libcasr::asan::AsanStacktrace;
+use libcasr::gdb::GdbStacktrace;
+use libcasr::go::GoStacktrace;
+use libcasr::java::JavaStacktrace;
+use libcasr::python::PythonStacktrace;
+use libcasr::stacktrace::ParseStacktrace;
+
+fuzz_target!(|data: &[u8]| {
```
Comment (SweetVishnya) [inline libcasr/fuzz/fuzz_targets/parse_stacktrace.rs:12]: Can we add fast fuzzing run to CI?

**Review by SweetVishnya — CHANGES_REQUESTED** [review]

File: libcasr/Cargo.toml (hunk 2 above)
Comment (anfedotoff) [reply libcasr/Cargo.toml:12]: We don't have tests for libcasr, it was copy-paste error :)

File: .github/workflows/main.yml [hunk 4]
```diff
@@ -32,6 +32,10 @@ jobs:
         export PATH=/root/.cargo/bin:$PATH
         cargo install cargo-fuzz
         cargo test --release --verbose
+    - name: Fuzz
+      run: |
+        cd libcasr
+        timeout 600 cargo +nightly fuzz parse_stacktrace
```
Comment (SweetVishnya) [inline .github/workflows/main.yml:38]: timeout may return non-zero exit code 124, we should check for both 0 and 124

Comment (SweetVishnya) [reply .github/workflows/main.yml:38]: also, we should explicitly verify that the crash directory is empty

**Review by SweetVishnya — CHANGES_REQUESTED** [review]

File: .github/workflows/main.yml (hunk 4 above)
Comment (SweetVishnya) [reply .github/workflows/main.yml:38]: we should also add some starting corpus for faster crash discovery

**Review by Avgor46 — APPROVED** [review]

**Comment by SweetVishnya** [comment]
Does cargo fuzz has non-zero exit code when crash is found?

**Comment by anfedotoff** [comment]
> Does cargo fuzz has non-zero exit code when crash is found?

Yes, I checked non-zero exit code is returned.

**Comment by SweetVishnya** [comment]
> > Does cargo fuzz has non-zero exit code when crash is found?
> 
> Yes, I checked non-zero exit code is returned.

The only thing left then is to collect some starting corpus.

File: libcasr/fuzz/fuzz_targets/parse_stacktrace.rs [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    if data.len() < 2 {
+        return;
+    }
+    let s = String::from_utf8_lossy(&data[1..]);
+    match data[0] % 5 {
+        0 => {
+            // Asan
+            if let Ok(st) = AsanStacktrace::extract_stacktrace(&s) {
+                let _ = AsanStacktrace::parse_stacktrace(&st);
```
Comment (hkctkuy) [inline libcasr/fuzz/fuzz_targets/parse_stacktrace.rs:21]: May be we need to add `crash_line()` func using?

Comment (SweetVishnya) [reply libcasr/fuzz/fuzz_targets/parse_stacktrace.rs:21]: Yeah, that would be nice

File: libcasr/fuzz/fuzz_targets/parse_stacktrace.rs [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+use libcasr::java::JavaStacktrace;
+use libcasr::python::PythonStacktrace;
+use libcasr::stacktrace::ParseStacktrace;
+
+fuzz_target!(|data: &[u8]| {
+    if data.len() < 2 {
+        return;
+    }
+    let s = String::from_utf8_lossy(&data[1..]);
+    match data[0] % 5 {
```
Comment (hkctkuy) [inline libcasr/fuzz/fuzz_targets/parse_stacktrace.rs:17]: May be we should split `parse_stacktrace.rs` to several separate ones for each `ParseStacktrace` impl?

**Review by hkctkuy — CHANGES_REQUESTED** [review]

File: libcasr/fuzz/fuzz_targets/parse_stacktrace.rs (hunk 6 above)
Comment (anfedotoff) [reply libcasr/fuzz/fuzz_targets/parse_stacktrace.rs:17]: The implementations of parsers are small. I think we could try this way first. If coverage is bad, we could separate them.

**Comment by SweetVishnya** [comment]
Can we move fuzzing to a separate pipeline? Thus, it can be run in parallel with tests. And CI will take less time.

**Review by SweetVishnya — APPROVED** [review]

**Pull request merged**